src/satellite/process.ts

```diff
@@ -301,38 +316,66 @@ export class SatelliteProcess implements Satellite {
   }
 
   // Perform a snapshot and notify which data actually changed.
-  async _performSnapshot(): Promise<void> {
-    const oplog = this.opts.oplogTable.toString()
-    const timestamp = new Date().toISOString()
+  async _performSnapshot(): Promise<Date> {
+    const timestamp = new Date()
 
-    const updateTimestamps = `
-      UPDATE ${oplog} set timestamp = ?
-        WHERE rowid in (
-          SELECT rowid FROM ${oplog}
-              WHERE timestamp is NULL
-              AND rowid > ?
-          ORDER BY rowid ASC
-          )
-    `
+    await this._updateOplogTimestamp(timestamp)
+    const oplogEntries = await this._getUpdatedEntries(timestamp)
 
-    const updateArgs = [timestamp, `${this._lastAckdRowId}`]
-    await this.adapter.run({ sql: updateTimestamps, args: updateArgs })
+    const promises: Promise<void | SatelliteError | RunResult>[] = []
 
-    const selectChanges = `
-      SELECT * FROM ${oplog} 
-      WHERE timestamp = ? ORDER BY rowid ASC
-    `
+    if (oplogEntries.length !== 0) {
+      promises.push(this._notifyChanges(oplogEntries))
 
-    const rows = await this.adapter.query({
-      sql: selectChanges,
-      args: [timestamp],
-    })
-    const results = rows as unknown as OplogEntry[]
+      const shadowEntries = new Map<string, ShadowEntry>()
+
+      for (const oplogEntry of oplogEntries) {
+        const [cached, shadowEntry] = await this._lookupCachedShadowEntry(
+          oplogEntry,
+          shadowEntries
+        )
 
-    const promises: Promise<void | SatelliteError>[] = []
+        /*
+        console.log("shadowentries:")
+        console.log(shadowEntries)
+        console.log("step:")
+        console.log(oplogEntry)
+        console.log(shadowEntry)
+        */
+
+        // Clear should not contain the tag for this timestamp, so if
+        // the entry was previously in cache - it means, that we already
+        // read it within the same snapshot
+        if (cached) {
+          oplogEntry.clearTags = encodeTags(
```

Comment:
I think this is okay, but let's just discuss to make sure we understand:
- The case when we end up with the tag in the clear set and in the tag set is INSERT.INSERT
- Is it a problem to have the tag in clear and tag sets? No, if when applying an incoming operation we first remove the clear set and then add the tag.

Having said that, I don't disagree with this code.